### PR TITLE
Modify to all_gather takes `group`

### PR DESCRIPTION
**Pull request by puhuk**
Description: Modify to all_gather takes `group`

Check list:

- [ ] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: tests/ignite/distributed/utils/__init__.py [hunk 1]
```diff
@@ -161,6 +161,17 @@ def _test_distrib_all_gather(device):
             idist.all_reduce([0, 1, 2])
 
 
+def _test_distrib_all_gather_group(device):
+
+    if idist.get_world_size() > 1:
+        rank = idist.get_rank()
+        group = [0, 1]
+
+        t = torch.tensor([rank], device=idist.device())
```
Comment (vfdev-5) [inline tests/ignite/distributed/utils/__init__.py:170]:
```suggestion
        t = torch.tensor([rank], device=device)
```

File: ignite/distributed/comp_models/native.py [hunk 2]
```diff
@@ -426,11 +426,13 @@ def _do_all_reduce(self, tensor: torch.Tensor, op: str = "SUM") -> torch.Tensor:
             dist.all_reduce(tensor, reduce_op)
             return tensor
 
-        def _do_all_gather(self, tensor: torch.Tensor) -> torch.Tensor:
+        def _do_all_gather(self, tensor: torch.Tensor, group: Optional[Union[Any, List[int]]] = None) -> torch.Tensor:
+            if group is not None and not isinstance(group, dist.ProcessGroup):
+                raise ValueError("Group should be list of int or ProcessGroup")
```
Comment (puhuk) [inline ignite/distributed/comp_models/native.py:431]: add test

File: ignite/distributed/utils.py [hunk 3]
```diff
@@ -354,7 +356,10 @@ def all_gather(tensor: Union[torch.Tensor, float, str]) -> Union[torch.Tensor, f
     if _need_to_sync and isinstance(_model, _SerialModel):
         sync(temporary=True)
 
-    return _model.all_gather(tensor)
+    if isinstance(group, list) and all(isinstance(item, int) for item in group):
```
Comment (puhuk) [inline ignite/distributed/utils.py:359]: loose the condition

File: ignite/distributed/comp_models/horovod.py [hunk 4]
```diff
@@ -185,7 +185,9 @@ def _do_manual_all_reduce(self, tensor: torch.Tensor, op: Any) -> torch.Tensor:
             # output can also torch min/max_return_type: (min/max_vals, indices)
             return reduced_res[0]
 
-        def _do_all_gather(self, tensor: torch.Tensor) -> torch.Tensor:
+        def _do_all_gather(self, tensor: torch.Tensor, group: Optional[Any] = None) -> torch.Tensor:
+            if group is not None:
+                raise NotImplementedError("all_reduce with group for horovod is not implemented")
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/distributed/comp_models/horovod.py:190]: And tests to update

**Review by vfdev-5 — APPROVED** [review]
Thanks @puhuk

**Pull request merged**